src/Integration/Persistence/FileBindingSerializer.cs

```diff
@@ -0,0 +1,230 @@
+
+            string configFile = this.GetFullConfigurationFilePath();
+            if (string.IsNullOrWhiteSpace(configFile))
+            {
+                return null;
+            }
+
+            sccFileSystem.QueueFileWrite(configFile, () =>
+            {
+                if (this.WriteBindingInformation(configFile, binding))
+                {
+                    return OnSuccessfulFileWrite(configFile);
+                }
+
+                return false;
+            });
+
+            if (this.Mode == WriteMode.Immediate && !sccFileSystem.WriteQueuedFiles())
+            {
+                return null; // write operation failed
+            }
+
+            return configFile;
+        }
+        
+        private BoundSonarQubeProject ReadBindingInformation(string configFile)
+        {
+            BoundSonarQubeProject bound = this.SafeDeserializeConfigFile(configFile);
+            if (bound?.ServerUri != null)
+            {
+                var credentials = this.credentialStore.ReadCredentials(bound.ServerUri);
+                if (credentials != null)
+                {
+                    bound.Credentials = new BasicAuthCredentials(credentials.Username,
+                        credentials.Password.ToSecureString());
+                }
+            }
+
+            return bound;
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Reliability",
+            "S3215:\"interface\" instances should not be cast to concrete types",
+            Justification = "Casting as BasicAuthCredentials is because it's the only credential type we support. Once we add more we need to think again on how to refactor the code to avoid this",
+            Scope = "member",
+            Target = "~M:SonarLint.VisualStudio.Integration.Persistence.FileBindingSerializer.WriteBindingInformation(System.String,SonarLint.VisualStudio.Integration.Persistence.BoundProject)~System.Boolean")]
+        private bool WriteBindingInformation(string configFile, BoundSonarQubeProject binding)
+        {
+            if (this.SafePerformFileSystemOperation(() => WriteConfig(configFile, binding)))
+            {
+                BasicAuthCredentials credentials = binding.Credentials as BasicAuthCredentials;
+                if (credentials != null)
+                {
+                    Debug.Assert(credentials.UserName != null, "User name is not expected to be null");
+                    Debug.Assert(credentials.Password != null, "Password name is not expected to be null");
+
+                    var creds = new Credential(credentials.UserName, credentials.Password.ToUnsecureString());
+                    this.credentialStore.WriteCredentials(binding.ServerUri, creds);
+                }
+                return true;
+            }
+
+            return false;
+        }
+
+        private void WriteConfig(string configFile, BoundSonarQubeProject binding)
+        {
+            string directory = Path.GetDirectoryName(configFile);
```

Comment:
Is it worth checking for null/empty before calling `GetDirectoryName`?

Reply:
With the code as it is at the moment, _WriteConfig_ will never be called if the the value is null/empty. I've added an assertion.

_WriteConfig_ is only every called from within _SafePerformFileSystemOperation_ so the call is inside a try-catch block. However, only specific exceptions are caught, and this doesn't include _ArgumentExceptions_.  I'll modify _SafePerformFileSystemOperation_ to catch all non-critical errors.